### PR TITLE
General Documentation for Cypress VSP - Cypress Resources Guide

### LINKED ISSUE #15832: General Documentation for use of Cypress on VSP
## Product Outline
N/A

## High Level User Story/ies
As a VFS team engineer, I need robust automated test framework documentation so I can create resilient and valuable tests for my VFS team product.

## Hypothesis or Bet
**If** _we make this change_ **then** _we expect this to happen_.
If we enhance the automated testing framework's documentation then engineers will have an easier time participating in test creation with Cypress on VA.gov.

## OKR
_Which Objective / Key Result does this epic push forward?_

50% of active VFS teams are participating in automated testing with Cypress.

### What must be true in order for you to consider this epic complete? 
- [x] General documentation for various areas of Cypress use have been created.
- [x] Existing documentation for use of Cypress on the platform has been enhanced.
- [x] Conduct an internal team peer review.
- [x] Submit a pull request

*Take into consideration Accessibility/QA needs as well as Product, Technical, and Design requirements.*

## How to configure this issue
- [x] **Labeled with Team** (`product support`, `analytics-insights`, `operations`, `tools-be`, `tools-fe`,`content-ia`, `service-design`, `vsa-*`)
- [x] **Labeled with Practice Area** (`backend`, `frontend`, `devops`, `design`, `research`, `product`, `ia`, `qa`, `analytics`, `contact center`, `research`, `accessibility`, `content`)

### PR DESCRIPTION
**Pull request by rjohnson2011**
Resolves https://github.com/department-of-veterans-affairs/va.gov-team/issues/15832

This PR creates a new markdown file, `vsp-cypress-resources`, which contains a table of contents and details regarding all custom written helpers and mocks actively used in Cypress tests. For convenience, a link to the file is placed within the `cypress-migration-guide` readme.

Each helper and mock is organized by category, linkable, and contains:

- The directory where the file can be found
- Import instructions
- A description of the purpose
- A reference guide for the objects properties
- Return value (if applicable)

### CODE REVIEW

File: platform/testing/end-to-end/vsp-cypress-resources.md [hunk 1]
```diff
@@ -0,0 +1,835 @@
...
+Import: `import { initMockProfile } from './cypress-helpers';`
+
+Purpose: Mocks the call for the profile.
+
+Reference:
+
+```javascript
+export const initMockProfile = (profile) => {
+  const facilityCode = profile.data.attributes.facility_code;
+  cy.route("GET", `/v0/gi/institutions/${facilityCode}`, profile).as(
```
Comment (1Copenut) [inline platform/testing/end-to-end/vsp-cypress-resources.md:146]: I'm guessing these might be API end points and this entire comment is off base. Please disregard if that's the case.

The heading for this section reads as though custom helpers are for the VAOS project only. This URL and several further down reference different projects, so I wasn't sure if the helpers are platform level (IE, available for any apps).

File: platform/testing/end-to-end/vsp-cypress-resources.md [hunk 2]
```diff
@@ -0,0 +1,835 @@
...
+    data: appointments,
+  };
+}
+```
+
+#### initCommunityCareMock <a name="init-community-care-mock"></a>
+
+Import: `import { initCommunityCareMock } from './vaos-cypress-helpers';`
+
+Purpose:
```
Comment (1Copenut) [inline platform/testing/end-to-end/vsp-cypress-resources.md:266]: Is there a defined purpose for this method?

**Review by 1Copenut — CHANGES_REQUESTED** [review]
4 comments. One's an actual change request, two are observations I need to log as a separate ticket, one's a question.

**Comment by 1Copenut** [comment]
@gmaciolek 🏓 for visibility.

**Comment by 1Copenut** [comment]
@rjohnson2011 Didn't know if you'd seen these content guides, but figured they're worth a mention:
* https://github.com/department-of-veterans-affairs/va.gov-team/blob/master/platform/content/content-best-practices.md
* https://design.va.gov/content-style-guide/

File: platform/testing/end-to-end/cypress-migration-guide.md [hunk 3]
```diff
@@ -42,6 +42,8 @@ vets-website
 - `plugins` contains custom tasks, which allow you to [tap into the Node environment](https://docs.cypress.io/guides/tooling/plugins-guide.html). This is valuable because all Cypress test code is executed in the browser, so plugins allow us to execute code in the Node process running outside of the browser.
 - `support` contains [custom Cypress commands](https://docs.cypress.io/api/cypress-api/custom-commands.html#Syntax). By default, the custom commands imported in `commands/index.js` are available to all of our Cypress tests. These commands can be invoked similarly to the built in commands. This feature is similar to [Nightwatch's custom commands](https://department-of-veterans-affairs.github.io/veteran-facing-services-tools/getting-started/common-tasks/new-end-to-end-test/#custom-nightwatch-commands). This is also where custom commands from Cypress plugins are imported, as can be seen in `index.js`.
 
+For a list of currently used custom written Cypress helpers and mocks for `vets-frontend` browse the [Cypress Resources Overview & Guide](vsp-cypress-resources.md).
```
Comment (pjhill) [inline platform/testing/end-to-end/cypress-migration-guide.md:45]: Is `vets-frontend` a reference to the front-end repo `vets-website`?
```suggestion
For a list of currently used custom written Cypress helpers and mocks for `vets-website` browse the [Cypress Resources Overview & Guide](vsp-cypress-resources.md).
```

File: platform/testing/end-to-end/vsp-cypress-resources.md [hunk 4]
```diff
@@ -0,0 +1,835 @@
...
+   9. [slots](#slots)
+   10. [getVAAppointmentMock](#get-va-appointment-mock)
+   11. [getExpressCareRequestCriteriaMock](#get-express-care-request-criteria-mock)
+   12. [getParentSiteMock](#get-parent-site-mock)
+
+# Cypress & VAOS Helperss <a name="cypress-vaos-helpers"></a>
+
+Custom helpers are reusable components for commonly needed test situations, like loading custom data, profiles, or appointments. Below is a list of helpers currently written along with import methods, purpose, and a reference guide.
+
+Consider writing a reusable helper when an action is needed to perform certain tests. They are located in `src/applications/gi/tests/e2e/cypress-helpers.js` and `src/applications/vaos/tests/e2e/vaos-cypress-helpers.js`.
```
Comment (pjhill) [inline platform/testing/end-to-end/vsp-cypress-resources.md:47]: These helpers are within the product folders for which the helpers are used. I'm not sure what the implication is for the scope of these helpers. Unless this is documented somewhere by Cypress that I can't find, I think we will need to experiment to see if Cypress loads these into the runner and permits their use throughout the testing suite.

At any rate, I think our recommendation should be slightly different than to encourage other product teams to be modifying the helper files located in a different team's product folder. Should each VFS team be creating a helper file in their product folder? Should we be consolidating these helper files into a larger file closer to the root?

**Review by pjhill — COMMENTED** [review]
Nice! Thanks for compiling this list!

Did you uncover any areas for potential enhancement? I noticed a few helpers with hardcoded values for example -- would there be an easy way to parameterize those?

**Comment by 1Copenut** [comment]
Created https://github.com/department-of-veterans-affairs/va.gov-team/issues/16338 for `hasFocusableCount` and `hasTabbableCount` updates.

File: platform/testing/end-to-end/vsp-cypress-resources.md [hunk 5]
```diff
@@ -0,0 +1,835 @@
...
+#### hasFocusableCount <a name="has-focusable-count"></a>
+
+Import: `import { hasFocusableCount } from './cypress-helpers'`
+
+Purpose: Checks if count of focusable elements is correct. Focusable elements are those in the normal tab order (native focusable elements or those with tabIndex 0). The count logic will break on tabindexes > 0 because we do not want to override the browser's base tab order.
+
+Reference:
+
+```javascript
+export const hasFocusableCount = (selector, count) => {
```
[the remark this answers is not in the document]
Comment (1Copenut) [reply platform/testing/end-to-end/vsp-cypress-resources.md:80]: https://github.com/department-of-veterans-affairs/va.gov-team/pull/16275#issuecomment-729752457

**Comment by gmaciolek** [comment]
Overall, these docs looks great to me. I assume these are VFS-facing docs? A few optional suggestions to consider:

1. Changing titles to:
- Nightwatch to Cypress Migration Guide
- Cypress Resources Reference Guide

2. At the top of the migration guide doc, add a little more detail/context or link to more information about the migration initiative. (e.g., Is there a doc that describes why we're migrating? Do we have expectations that teams should migrate by a certain date? What is the implication if teams don't migrate?)

3. At the top of the resources doc, if we're going to include the bit about "as of 11/2020" I suggest also adding information about whether there are plans to add additional resources (and when). If we don't know, or if there are no plans to add more--I'd leave the date off. (Otherwise, if someone looks at it in the spring, for example, they might wonder whether it's still accurate.)

Thanks for looping me in @1Copenut !

File: platform/testing/end-to-end/vsp-cypress-resources.md [hunk 6]
```diff
@@ -0,0 +1,835 @@
...
+   5. [supportedSites](#supported-sites)
+   6. [facilities](#facilities)
+   7. [facilities983](#facilities-983)
+   8. [clinicList983](#clinic-list-983)
+   9. [slots](#slots)
+   10. [getVAAppointmentMock](#get-va-appointment-mock)
+   11. [getExpressCareRequestCriteriaMock](#get-express-care-request-criteria-mock)
+   12. [getParentSiteMock](#get-parent-site-mock)
+
+# Cypress & VAOS Helperss <a name="cypress-vaos-helpers"></a>
```
[the remark this answers is not in the document]
Comment (rjohnson2011) [reply platform/testing/end-to-end/vsp-cypress-resources.md:43]: Thanks for catching that!

File: platform/testing/end-to-end/cypress-migration-guide.md (hunk 3 above)
Comment (rjohnson2011) [reply platform/testing/end-to-end/cypress-migration-guide.md:45]: Yes! Thanks for noticing that.

File: platform/testing/end-to-end/vsp-cypress-resources.md (hunk 4 above)
Comment (rjohnson2011) [reply platform/testing/end-to-end/vsp-cypress-resources.md:47]: Solid points to consider. Seems when the tests were more localized to a particular team, it made sense to store them in that directory. Now that we're publicizing their contacts (and reusability) there's a strong argument for these to live in a parent directory. The caveat to that would be tests that are scoped and only usable within a particular directory.

Whatever way we go, DRY principles ought to be encouraged. Let's avoid having the same commonly used helper methods in various directories throughout the app.

File: platform/testing/end-to-end/vsp-cypress-resources.md (hunk 2 above)
Comment (rjohnson2011) [reply platform/testing/end-to-end/vsp-cypress-resources.md]: Added.

**Review by pjhill — APPROVED** [review]
LGTM -- as discussed in the stand up this morning, any effort to consolidate/move helper functions closer to the root of the repo instead of scattering them amongst individual product folders can be worked under a separate effort.

**Pull request merged**